composer/algorithms/cutmix/cutmix.py

```diff
@@ -210,48 +193,103 @@ class CutMix(Algorithm):
+        interpolate_loss: bool = False,
         uniform_sampling: bool = False,
         input_key: Union[str, int, Tuple[Callable, Callable], Any] = 0,
         target_key: Union[str, int, Tuple[Callable, Callable], Any] = 1,
     ):
-        self.num_classes = num_classes
         self.alpha = alpha
+        self.interpolate_loss = interpolate_loss
         self._uniform_sampling = uniform_sampling
+
         self._indices = torch.Tensor()
         self._cutmix_lambda = 0.0
         self._bbox: Tuple[int, int, int, int] = (0, 0, 0, 0)
+        self._permuted_target = torch.Tensor()
+        self._adjusted_lambda = 0.0
         self.input_key, self.target_key = input_key, target_key
 
     def match(self, event: Event, state: State) -> bool:
-        return event == Event.AFTER_DATALOADER
+        if self.interpolate_loss:
+            return event in [Event.BEFORE_FORWARD, Event.BEFORE_BACKWARD]
+        else:
+            return event in [Event.BEFORE_FORWARD, Event.BEFORE_LOSS]
 
     def apply(self, event: Event, state: State, logger: Logger) -> None:
         input = state.batch_get_item(key=self.input_key)
         target = state.batch_get_item(key=self.target_key)
 
-        assert isinstance(input, Tensor) and isinstance(target, Tensor), \
-            'Multiple tensors for inputs or targets not supported yet.'
-        alpha = self.alpha
-
-        # these are saved only for testing
-        self._indices = _gen_indices(input)
-        _cutmix_lambda = _gen_cutmix_coef(alpha)
-        self._bbox = _rand_bbox(input.shape[2], input.shape[3], _cutmix_lambda, uniform_sampling=self._uniform_sampling)
-        self._cutmix_lambda = _adjust_lambda(_cutmix_lambda, input, self._bbox)
+        if not isinstance(input, torch.Tensor):
+            raise NotImplementedError('Multiple tensors for inputs not supported yet.')
+        if not isinstance(target, torch.Tensor):
+            raise NotImplementedError('Multiple tensors for targets not supported yet.')
 
-        new_input, new_target = cutmix_batch(
-            input=input,
-            target=target,
-            num_classes=self.num_classes,
-            alpha=alpha,
-            bbox=self._bbox,
-            indices=self._indices,
-        )
+        alpha = self.alpha
 
-        state.batch_set_item(key=self.input_key, value=new_input)
-        state.batch_set_item(key=self.target_key, value=new_target)
+        if event == Event.BEFORE_FORWARD:
+            # these are saved only for testing
+            self._indices = _gen_indices(input)
+
+            _cutmix_lambda = _gen_cutmix_coef(alpha)
+            self._bbox = _rand_bbox(input.shape[2],
+                                    input.shape[3],
+                                    _cutmix_lambda,
+                                    uniform_sampling=self._uniform_sampling)
+            self._adjusted_lambda = _adjust_lambda(input, self._bbox)
+
+            new_input, self._permuted_target, _, _ = cutmix_batch(input=input,
```

Comment:
do we need `cutmix_batch` to return `bbox`? I don't see it being used anywhere in the code or examples

Reply:
I think it needs to return `bbox` such that users of the functional interface can replicate what cutmix does to the target tensors when they have spatial dimensions. The composer algorithm implementation handles this automatically, but it would have to be done manually with the current functional implementation.